ui/src/app/utils/runtime-hooks.spec.tsx

```diff
@@ -0,0 +1,286 @@
+  });
+
+  it('should delete runtime and create a new one when disk is decreased', async () => {
+    const newDisk = null;
+    const [, setRequest] = testUseCustomRuntime();
+
+    const newRuntime = defaultRuntime();
+    newRuntime.gceConfig.diskSize = currentRuntime.gceConfig.diskSize / 2;
+
+    await act(async () => {
+      setRequest({ runtime: newRuntime, detachedDisk: newDisk });
+    });
+
+    // Triggers a delete, because disk size decreased
+    await waitFor(() => {
+      expect(deleteRuntimeSpy).toHaveBeenCalled();
+    });
+
+    await waitFor(() => {
+      expect(initializerSpy).toHaveBeenCalled();
+    });
+  });
+
+  it('should delete runtime and create a new one when current runtime is in error state', async () => {
+    const [, setRequest] = testUseCustomRuntime();
+
+    // Set the new runtime to be the same as the current runtime, but
+    // by using the spread operator, we create a new object
+    // rather than a reference
+    const newRuntime = { ...currentRuntime };
+
+    // This means that settign the status to error will not affect the new runtime.
+    currentRuntime.status = RuntimeStatus.ERROR;
+
+    await act(async () => {
+      setRequest({ runtime: newRuntime, detachedDisk: null });
+    });
+
+    await waitFor(() => {
+      expect(deleteRuntimeSpy).toHaveBeenCalled();
+    });
+
+    await waitFor(() => {
+      expect(initializerSpy).toHaveBeenCalled();
+    });
+  });
+
+  it('should allow Dataproc -> PD transition', async () => {
+    currentRuntime = defaultDataProcRuntime();
+    runtimeStore.set({
+      workspaceNamespace: workspaceDataStub.namespace,
+      runtime: currentRuntime,
+      runtimeLoaded: true,
+    });
+
+    const [, setRequest] = testUseCustomRuntime();
+
+    const newRuntime = defaultGceRuntime();
+
+    await act(async () => {
+      setRequest({ runtime: newRuntime, detachedDisk: null });
+    });
+
+    await waitFor(() => {
+      expect(deleteRuntimeSpy).toHaveBeenCalled();
+    });
+
+    await waitFor(() => {
+      expect(initializerSpy).toHaveBeenCalled();
+    });
+  });
+
+  it("should update runtime, and not delete, when dataproc's master disk size is increased", async () => {
+    currentRuntime = {
+      ...defaultRuntime(),
+      status: RuntimeStatus.RUNNING,
+      configurationType: RuntimeConfigurationType.USER_OVERRIDE,
+      gceConfig: null,
+      gceWithPdConfig: null,
+      dataprocConfig: {
+        masterMachineType: 'n1-standard-4',
+        masterDiskSize: 1000,
+        numberOfWorkers: 2,
+        numberOfPreemptibleWorkers: 0,
+        workerMachineType: 'n1-standard-4',
+        workerDiskSize: DATAPROC_MIN_DISK_SIZE_GB,
+      },
+    };
+
+    runtimeStore.set({
+      workspaceNamespace: workspaceDataStub.namespace,
+      runtime: currentRuntime,
+      runtimeLoaded: true,
+    });
+
+    const [, setRequest] = testUseCustomRuntime();
+    const newRuntime = JSON.parse(JSON.stringify(currentRuntime));
```

Comment:
interesting approach!  I recall you used `{ ...whatever }` to copy, in another case.